### PR TITLE
Add child client factory methods

### PR DESCRIPTION
**Pull request by jorgerangel-msft**
This PR adds the child client factory methods for clients.

fixes: https://github.com/microsoft/typespec/issues/3690

### CODE REVIEW

File: packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/Unbranded-TypeSpec.tsp [hunk 1]
```diff
@@ -366,3 +366,31 @@ op stillConvenient(): void;
 @head
 @convenientAPI(true)
 op headAsBoolean(@path id: string): void;
+
+@route("/audio")
```
Comment (m-nash) [inline packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/Unbranded-TypeSpec.tsp:370]: lets convert these to unit tests only, cadl ranch will cover these for generated shape.

**Review by ArcturusZhang — COMMENTED** [review]
LGTM

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/test/Providers/ClientProviders/ClientProviderSubClientTests.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                    clients: () => [_animalClient, _dogClient, _catClient, _huskyClient]);
+        }
+
+        // This test validates that the generated code is correct when the client has one direct subclient.
+        [Test]
+        public void ServiceClientWithSubClient()
+        {
+            var client = new InputClient(TestClientName, "TestClient description", [], [], null);
+            var clientProvider = new ClientProvider(client);
+            var writer = new TypeProviderWriter(clientProvider);
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/test/Providers/ClientProviders/ClientProviderSubClientTests.cs:33]: Instead of using TypeProviderWriter which writes the entire file, this test is only concerned with the sub client methods on the client, lets isolate and call codeWriter.WriteMethod so that is the only thing that needs to be in the test data file and we don't have changes to this file if something unrelated changes like the fields in a client.

**Review by m-nash — APPROVED** [review]
